### ISSUE TITLE
Serving large csv

### ISSUE THREAD

**Issue by mtog** [1001genomes/AraGWAS#129]
For larger files (>5K associations), we need to serve them directly to the client via StreamingHttpResponse similar to https://docs.djangoproject.com/en/1.11/howto/outputting-csv/#streaming-large-csv-files

**Comment by timeu** [comment]
We are already using StreamingHttpResponse (see [here](https://github.com/1001genomes/AraGWAS/blob/master/aragwas_server/gwasdb/rest.py#L552))

**Comment by mtog** [comment]
Yes, but we do it once the file is fully generated, in the example given on the django website they use the csv writer to directly write the file in download chunks.

Le 29 oct. 2017 à 17:03, Ümit Seren <notifications@github.com<mailto:notifications@github.com>> a écrit :


We are already using StreamingHttpResponse (see here<https://github.com/1001genomes/AraGWAS/blob/master/aragwas_server/gwasdb/rest.py#L552>)

**Comment by timeu** [comment]
Ok I understand, but then we need to modify to https://github.com/1001genomes/AraGWAS/blob/master/aragwas_server/gwasdb/es2csv.py to directly generate the csv instead of first generating the json temporary file and then generating the csv file. Additionally we have to do this in the export folder and use some unique identifier so that it doesn't collide when two users download the file at the same time .

**Comment by mtog** [comment]
Yes. That's right, I didn't think about two user downloads collision.
Another solution would be to store the generated files for a longer time and save their name together with a list of filters used for something like 30 days. This would allow not to re-generate the large files every time.

**Comment by timeu** [comment]
Yes but then we still can run into a race condition when two users want to download the top associations with the same filter, because then you would start to overwrite the file by both processes. 
We would have to store the information that a file is currently being created and then somehow allow the second user to stream from the same file that is being created by the request of the first user

**Comment by mtog** [comment]
I see. Is this automatically done if using a webserver such as nginx for serving files?

**Comment by timeu** [comment]
no not really, but we could think about adding a cache layer (redis) or directly using django cache to cache the generated files. 
It won't solve the race condition, but at least once the file is generated it can be cached and subsequent calls (after it has been cached) are served from that.
